clang/include/clang/Sema/Sema.h

```diff
@@ -13081,7 +13081,7 @@ class Sema final {
   /// Tells whether given variable is a SYCL explicit SIMD extension's "private
   /// global" variable - global variable in the private address space.
   bool isSYCLEsimdPrivateGlobal(VarDecl *VDecl) {
-    return getLangOpts().SYCLIsDevice && getLangOpts().SYCLExplicitSIMD &&
+    return getLangOpts().SYCLIsDevice && VDecl->hasAttr<SYCLSimdAttr>() &&
```

Comment:
Is this going to potentially break ABI for people? They used to enable a flag at compile time and now that flag does nothing in this case unless the user changes their variable declarations.

Reply:
> They used to enable a flag at compile time and now that flag does nothing in this case unless the user changes their variable declarations.

Yes. Now we enforce users to mark ESIMD globals with the sycl_explicit_simd attribute. I'm now working on the patch that will forbid using ESIMD globals in the regular SYCL context.

> Is this going to potentially break ABI for people?

I can't see how it can break ABI. Can you please describe a scenario you have in mind?

Reply:
Keep in mind that I don't know this ESIMD functionality at all, so maybe I have unrealistic concerns.

I'm worried that there's a global variable in TU1 which is compiled without SYCLExplicitSIMD enabled before landing this patch. In that TU, the variable is not private global. In TU2, there's a global variable with the same name as the one from TU1, but TU2 is compiled SYCLExplicitSIMD and so that variable was private global, so there was (presumably) not linkage conflicts with TU1. The user updates the compiler, recompiles TU2 without changes, and now that global variable is no longer private global. When they go to link against TU1 without recompiling it, that seems like it could do bad things.

Reply:
That's fine. Thanks for bringing that.

I think that the scenario you mentioned is not yet officially supported. We don't support having TUs compiled with different SYCLExplicitSIMD flags in the same program (aka mixing regular SYCL and ESIMD code in the same program).

Then, if the user has ESIMD private global in TU2 marked with `ESIMD_PRIVATE`, then they would be safe after updating the compiler since this will help disambiguate globals in TU1 and TU2.

Reply:
Well, the same issue presents itself if both TUs are compiled with the same SYCLExplicitSIMD flags, too.

TU1 and TU2 are compiled with the flag enabled and a global variable named `x`, so each TU has their own private global variable. The user downloads a new version of DPC++, recompiles both TUs (still with the flag enabled), now they get link errors because the two `x` variables are no longer private global (because neither variable was marked with the attribute).

Reply:
I think this is a valid concern. But ESIMD still has experimental status, so we don't expect much legacy code. If there is users are expected to be tolerant due to the experimental status.
Nevertheless, I think we need a test for ESIMD-specific ABI compatibility - the scenario @AaronBallman describes above. Basically, the test would contain .ll produced by fixed (current) version, plus .cpp compiled each time test runs. Then the .ll is linked  with what compilation produces. @DenisBakhvalov - this PR looks like a good place to add it, or should we do this as a separate PR, what do you think?

Reply:
> I think this is a valid concern. But ESIMD still has experimental status, so we don't expect much legacy code. If there is users are expected to be tolerant due to the experimental status.

Good to know, and that seems reasonable to me.